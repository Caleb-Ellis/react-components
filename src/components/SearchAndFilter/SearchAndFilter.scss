@import "~vanilla-framework/scss/base_placeholders";

.search-and-filter {
  border: 1px solid $color-mid-dark;
  position: relative;

  &__search-container {
    align-items: center;
    display: flex;
    flex-wrap: wrap;
    margin: 0;
    min-height: 40px;
    overflow: hidden;
    padding: 0 2rem 0 0.5rem;
    position: relative;

    // if search container is not active
    &[data-active="false"] {
      // hide the overflowing input field
      [data-overflowing="true"] {
        display: none;
      }
    }

    .p-chip {
      margin-bottom: 0.5rem;
    }

    .search-and-filter__selected-count {
      color: $color-information;
      cursor: pointer;
      position: absolute;
      right: 0.5rem;
      top: 6px;
    }

    &[aria-expanded="false"] {
      .p-chip {
        margin-bottom: 0.5rem;
      }
    }

    &[aria-expanded="true"] {
      max-height: 100%;

      .search-and-filter__selected-count {
        display: none;
      }
    }

    .p-search-box {
      flex-grow: 1;
    }
  }

  .p-icon--search {
    display: none;
  }

  [class*="p-contextual-menu"],
  .p-contextual-menu__dropdown {
    display: block;
    max-width: 100%;
    width: 100%;
  }

  .p-contextual-menu__dropdown {
    left: 0;
    padding: 1rem;
    top: 0;
  }

  &__panel {
    padding: 0;
    transition: opacity 0.25s;

    &[aria-hidden="true"] {
      opacity: 0;
      pointer-events: none;
    }
  }

  .search-prompt {
    background: $color-light;
    cursor: pointer;
    margin-left: -1rem;
    margin-right: -1rem;
    overflow: hidden;
    padding: 1rem;
    text-overflow: ellipsis;
    white-space: nowrap;
  }

  &__clear {
    border: none;
    height: 2rem;
    margin: 0;
    max-width: 48px;
    position: absolute;
    right: 3px;
    top: 3px;
    z-index: 999;
  }
}
