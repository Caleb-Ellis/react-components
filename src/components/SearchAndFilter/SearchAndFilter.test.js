import { shallow, mount } from "enzyme";
import React from "react";

import SearchAndFilter from "./SearchAndFilter";

const sampleData = [
  {
    id: 0,
    heading: "Clouds",
    chips: [{ value: "Google" }, { value: "AWS" }, { value: "Azure" }],
  },
  {
    id: 1,
    heading: "Regions",
    chips: [
      { value: "us-east1" },
      { value: "us-east2" },
      { value: "us-east3" },
    ],
  },
  {
    id: 3,
    heading: "Owner",
    chips: [{ value: "foo" }, { value: "bar" }, { value: "baz" }],
  },
];

describe("Search and filter", () => {
  it("renders", () => {
    const wrapper = shallow(<SearchAndFilter />);
    expect(wrapper).toMatchSnapshot();
  });

  it("hide the clear button when there is no value in search box", () => {
    const wrapper = mount(<SearchAndFilter />);
    expect(wrapper.find(".p-search-box__reset").exists()).toBe(false);
  });

  it("shows panel on focus", () => {
    const mockOnChange = jest.fn();
    const wrapper = mount(
      <SearchAndFilter
        externallyControlled
        onChange={mockOnChange}
        filterPanelData={sampleData}
      />
    );
    expect(
      wrapper.find(".search-and-filter__panel").prop("aria-hidden")
    ).toEqual(true);
    wrapper.find(".p-search-box__input").simulate("focus");
    expect(
      wrapper.find(".search-and-filter__panel").prop("aria-hidden")
    ).toEqual(false);
  });

  it("shows panel on click", () => {
    const mockOnChange = jest.fn();
    const wrapper = mount(
      <SearchAndFilter
        externallyControlled
        onChange={mockOnChange}
        filterPanelData={sampleData}
      />
    );
    expect(
      wrapper.find(".search-and-filter__panel").prop("aria-hidden")
    ).toEqual(true);
    wrapper.find(".p-search-box__input").simulate("click");
    expect(
      wrapper.find(".search-and-filter__panel").prop("aria-hidden")
    ).toEqual(false);
  });

  it("should hide chip overflow counter when none overflow", () => {
    const mockOnChange = jest.fn();
    // Jest is unaware of layout so we must mock the offsetTop and offsetHeight
    // of the chips
    Object.defineProperty(HTMLElement.prototype, "offsetHeight", {
      configurable: true,
      value: 40,
    });
    Object.defineProperty(HTMLElement.prototype, "offsetTop", {
      configurable: true,
      value: 40,
    });
    const wrapper = mount(
      <SearchAndFilter
        externallyControlled
        onChange={mockOnChange}
        filterPanelData={sampleData}
      />
    );
    expect(wrapper.find(".search-and-filter__selected-count").length).toEqual(
      0
    );
  });

  it("show overflow chip counter when chips overflow", () => {
    // Jest is unaware of layout so we must mock the offsetTop and offsetHeight
    // of the chips
    Object.defineProperty(HTMLElement.prototype, "offsetHeight", {
      configurable: true,
      value: 40,
    });
    Object.defineProperty(HTMLElement.prototype, "offsetTop", {
      configurable: true,
      value: 100,
    });
    const wrapper = mount(
      <SearchAndFilter externallyControlled filterPanelData={sampleData} />
    );
    wrapper
      .find(".filter-panel-section__chips .p-chip")
      .first()
      .simulate("click");
    expect(wrapper.find(".search-and-filter__selected-count").text()).toEqual(
      "+1"
    );
  });

  it("all chips are shown when counter is clicked", () => {
    // Jest is unaware of layout so we must mock the offsetTop and offsetHeight
    // of the chips
    Object.defineProperty(HTMLElement.prototype, "offsetHeight", {
      configurable: true,
      value: 40,
    });
    Object.defineProperty(HTMLElement.prototype, "offsetTop", {
      configurable: true,
      value: 100,
    });
    const wrapper = mount(
      <SearchAndFilter externallyControlled filterPanelData={sampleData} />
    );
    expect(
      wrapper.find(".search-and-filter__search-container").prop("aria-expanded")
    ).toEqual(false);
    wrapper
      .find(".filter-panel-section__chips .p-chip")
      .first()
      .simulate("click");
    wrapper.find(".search-and-filter__selected-count").simulate("click");
    expect(
      wrapper.find(".search-and-filter__search-container").prop("aria-expanded")
    ).toEqual(true);
  });

  it("search prompt appears when search field has search term", () => {
    const wrapper = mount(
      <SearchAndFilter externallyControlled filterPanelData={sampleData} />
    );
    expect(wrapper.find(".search-prompt").length).toEqual(0);
    wrapper
      .find(".p-search-box__input")
      .simulate("change", { target: { value: "My new value" } });
    expect(wrapper.find(".search-prompt").length).toEqual(1);
  });

  it("no search results appear for unknown search term", () => {
    const wrapper = mount(
      <SearchAndFilter externallyControlled filterPanelData={sampleData} />
    );
    expect(wrapper.find(".filter-panel-section").length).toEqual(3);
    wrapper
      .find(".p-search-box__input")
      .simulate("change", { target: { value: "Unknown value" } });
    expect(wrapper.find(".filter-panel-section").length).toEqual(0);
  });

  it("correct number of panels appear for matching search terms", () => {
    const wrapper = mount(
      <SearchAndFilter externallyControlled filterPanelData={sampleData} />
    );
    wrapper
      .find(".p-search-box__input")
      .simulate("change", { target: { value: "Google" } });
    expect(wrapper.find(".filter-panel-section").length).toEqual(1);
    wrapper
      .find(".p-search-box__input")
      .simulate("change", { target: { value: "re" } });
    expect(wrapper.find(".filter-panel-section").length).toEqual(2);
  });

  it("Matching search terms are highlighted with strong tag", () => {
    const wrapper = mount(
      <SearchAndFilter externallyControlled filterPanelData={sampleData} />
    );
    wrapper
      .find(".p-search-box__input")
      .simulate("change", { target: { value: "Google" } });
    const firstChip = wrapper.find(".p-chip").first();
    expect(firstChip.find(".p-chip__value").html()).toEqual(
      '<span class="p-chip__value"><strong>Google</strong></span>'
    );
  });
});
