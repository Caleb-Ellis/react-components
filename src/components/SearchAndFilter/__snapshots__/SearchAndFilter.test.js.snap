// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search and filter renders 1`] = `
<div
  className="search-and-filter"
>
  <div
    aria-expanded={false}
    className="search-and-filter__search-container"
    data-active={false}
  >
    <ForwardRef
      autocomplete="off"
      data-overflowing={false}
      externallyControlled={false}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onSubmit={[Function]}
      placeholder="Search and filter"
      value=""
    />
  </div>
</div>
`;
